grading-vm/orca_grader/__main__.py

```diff
@@ -82,25 +89,33 @@ def push_results_with_exception(job_json_string: str, e: Exception):
 def run_local_job(job_path: str, no_container: bool, container_command: Optional[List[str]]):
   retriever = LocalGradingJobRetriever(job_path)
   job_string = retriever.retrieve_grading_job()
-  if not can_execute_job(job_string):
-    push_results_with_exception(job_string, InvalidGradingJobJSONException())
-  else:
-    run_grading_job(job_string, no_container, container_command)
+  run_grading_job(job_string, no_container, container_command)
   clean_up_unused_images()
 
 def process_redis_jobs(redis_url: str, no_container: bool, container_command: List[str] | None):
   retriever = RedisGradingJobRetriever(redis_url)
-  while True: 
-    job_string = retriever.retrieve_grading_job()
-    if not can_execute_job(job_string):
-      push_results_with_exception(job_string, InvalidGradingJobJSONException())
-      continue
-    run_grading_job(job_string, no_container, container_command)
-    if APP_CONFIG.enable_diagnostics:
-      log_diagnostics(get_redis_client(redis_url), 
-                      time.time_ns(), 
-                      json.loads(job_string)["key"],
-                      JobState.COMPLETED)
+  killer = GracefulKiller()
+  with concurrent.futures.ThreadPoolExecutor(max_workers=2) as futures_executor:
+    stop_future = futures_executor.submit(killer.wait_for_stop_signal)
+    while True:
+      job_string = retriever.retrieve_grading_job()
+      job_future = futures_executor.submit(run_grading_job, job_string, no_container, container_command)
+      done, not_done = concurrent.futures.wait([stop_future, job_future], return_when="FIRST_COMPLETED")
+      if stop_future in done and job_future in not_done:
```

Comment:
Make explicit the three future states: both are done, stop only is done, job only is done.